File: Napakalaki Ruby/lib/player.rb
# By: Miguel Robles Urquiza
#     Jesús Sánchez de Lechina Tejada
  require_relative "treasure"
  require_relative "dice"
  require_relative "bad_consequence"
  require_relative "treasure_kind"
  require_relative "combat_result"

module NapakalakiGame


  class Player
    @@MAXLEVEL = 10
    @@INITIALLEVEL = 1

    def initialize(name)

      @name = name
      @level = @@INITIALLEVEL
      @dead = true
      @canISteal = true
      @hiddenTreasures = Array.new
      @visibleTreasures = Array.new
      @pendingBadConsequence = nil
      @enemy = nil
    end
    
    def newCopy(p)
      np = new(p.getName)
      np.setVT Array.new p.getVT
      np.setHT Array.new p.getHT
      np.setLevel p.getLevels
      np.setDead p.isDead
      np.setCanI p.getCanI
      np.setEnemy p.getEnemy
      np.setPendingBadConsequence p.getPendingBadConsequence
      
      np
    end
    
    def self.MAXLEVEL
      @@MAXLEVEL
    end
    
    def to_s
      c_level = getCombatLevel
      "#{@name} \nNivel del jugador: #{@level}\nNivel de combate: #{c_level}"
    end
    
    def bringToLife
      @dead = false
    end

    def getCombatLevel
      clevel = @level
      @visibleTreasures.each { |trea| clevel += trea.getBonus }
      return clevel
    end
    
    def getOponentLevel m 
      m.getCombatLevel
    end
    
    def shouldConvert
     
      return true if(Dice.instance.nextNumber == 6)
      return false
    end
    

    def incrementLevels(i)
      if ((@level+i)>10)
        @level = 10
      else
        @level += i
      end    
    end

    def decrementLevels(i)
      if (i>=@level)
        @level = 1
      else
        @level -= i
      end
    end

 
    
    def applyPrize(m)
      incrementLevels(m.getLevelsGained)
      if (m.getTreasuresGained > 0)
        dealer = CardDealer.instance
        m.getTreasuresGained.times do 
          treasure = dealer.nextTreasure
          @hiddenTreasures << treasure
        end
      end
    end

    def applyBadConsequence(m)

      badConsequence = m.getBadConsequence
      nLevels = badConsequence.getLevels
      decrementLevels(nLevels)
      pendingBad = badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
      setPendingBadConsequence(pendingBad)
    end   
    
    def howManyVisibleTreasures(tKind)
      i = 0
      @visibleTreasures.each do |t|
        if (t.getType == tKind)
          i += 1
        end
      end 
      i
    end
    
    def dieIfNoTreasures
      if @visibleTreasures.empty? and @hiddenTreasures.empty?
      @dead = true
      end
    end
    
    def validState
      @hiddenTreasures.length <= 4 && (@pendingBadConsequence == nil || @pendingBadConsequence.isEmpty)      
    end 
    
    def isDead
      @dead
    end     
   
    def canISteal
      @canISteal
    end
      
    
    def getName
      @name
    end
     
    def getHiddenTreasures
      @hiddenTreasures
    end

    def getVisibleTreasures
      @visibleTreasures
    end

    def getLevels
      @level
    end
    
    def getEnemy
      @enemy
    end
    
    def getPendingBadConsequence
      @pendingBadConsequence
    end
    
    def setEnemy(enemy)
      @enemy = enemy
    end
    
    def setVT(vt)
      @visibleTreasures = vt
    end
    
    def setHT(ht)
      @hiddenTreasures = ht
    end
    
    def setPendingBadConsequence(b)
      @pendingBadConsequence = b
    end
    

    def canYouGiveMeATreasure
      return true if (@hiddenTreasures.size() > 0)
      return false
    end

    def haveStolen
      @canISteal = false
    end

   def canMakeTreasureVisible(t)
      
      if (t.getType != TreasureKind::ONEHAND && t.getType != TreasureKind::BOTHHANDS)
          return false if (treasureInUse(t.getType))
      return true
      else                                          
        una_mano = 0
        dos_manos = 0
        @visibleTreasures.each do |i|
          puts i
          una_mano += 1 if (TreasureKind::ONEHAND == i.getType)
          dos_manos += 1 if (TreasureKind::BOTHHANDS == i.getType)

        end
        if (una_mano == 0 && dos_manos == 0)                        
          return true
        end
      end
      return una_mano == 1 && t.getType() == TreasureKind::ONEHAND && dos_manos == 0;
    end
        
    def treasureInUse (type)
      @visibleTreasures.each do |i|
        if(type == i.getType)
          return true
        end
      end
      return false
    end

   def combat(m)
      myLevel = getCombatLevel
      monsterLevel = getOponentLevel m
      if (@canISteal == false)
        dice = Dice.instance
        number = dice.nextNumber
        if (number < 3)
          enemyLevel = @enemy.getCombatLevel
          monsterLevel += enemyLevel
        end
      end
      if (myLevel > monsterLevel)
        applyPrize(m)
        if (@level >= @@MAXLEVEL)
          combatResult = CombatResult::WINGAME
        else
          combatResult = CombatResult::WIN
        end
      else 
        applyBadConsequence(m)
        if(shouldConvert)
          combatResult = CombatResult::LOSEANDCONVERT
        else
          combatResult = CombatResult::LOSE
        end     
      end
      combatResult
    end
    
   def makeTreasureVisible(t)
      canI = canMakeTreasureVisible(t)
      if (canI)
        @visibleTreasures << t
        @hiddenTreasures.delete(t)
      end
    end

    def discardVisibleTreasure(t)
      @visibleTreasures.delete(t)
      if (@pendingBadConsequence != nil && @pendingBadConsequence.isEmpty() == false)
        @pendingBadConsequence.substractVisibleTreasure t
        if @pendingBadConsequence.getNVisibleTreasures == 0 && @pendingBadConsequence.getNHiddenTreasures == 0
          @pendingBadConsequence = nil
        end
      end
      
      dieIfNoTreasures
    end

    def discardHiddenTreasure(t)
      @hiddenTreasures.delete(t)
      if (@pendingBadConsequence != nil && @pendingBadConsequence.isEmpty() == false)
          @pendingBadConsequence.substractHiddenTreasure t
          if @pendingBadConsequence.nHiddenTreasures == 0 && @pendingBadConsequence.nVisibleTreasures != 0
          @pendingBadConsequence = nil
          end
      end
      dieIfNoTreasures
    end

    def initTreasures
      dealer = CardDealer.instance
      dice = Dice.instance
      bringToLife
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      number = dice.nextNumber
      if (number > 1)
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure        
      end
      if (number == 6)
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end

    def stealTreasure
      canI = canISteal
      if (canI)
        if (@enemy.canYouGiveMeATreasure)
          treasure = @enemy.giveMeATreasure
          @hiddenTreasures << treasure
          haveStolen
        end
      else
        treasure = nil
      end
      return treasure
    end

    def giveMeATreasure
      indice = rand(@hiddenTreasures.size)
      tr = @hiddenTreasures[indice]
      @hiddenTreasures.delete_at(indice)
      return tr
    end

    def discardAllTreasures
      cVisible= Array.new(@visibleTreasures)
      cHidden = Array.new(@hiddenTreasures)
      cVisible.each do |treasure|
        discardVisibleTreasure(treasure)
      end
      cHidden.each do |treasure|
        discardHiddenTreasure(treasure)
      end
    end
    
  private :bringToLife, :incrementLevels, :decrementLevels, :setPendingBadConsequence, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures, :haveStolen
  protected :canYouGiveMeATreasure, :giveMeATreasure, :getCombatLevel, :getName,:getLevels,  :getPendingBadConsequence
  end

end

File: Napakalaki Java/src/NapakalakiGame/Player.java
package NapakalakiGame;

import java.util.ArrayList;
import GUI.Dice;

/**
 *
 * @author Jesús Sánchez de Lechina Tejada & Miguel Ángel Robles Urquiza
 */
public class Player {
    static final int INITIALLEVEL = 1;
    static final int MAXLEVEL = 10;


    private String name = null;
    protected int level;
    protected boolean dead = true;
    protected boolean canISteal = true;
    protected Player enemy;
    protected BadConsequence pendingBadConsequence;

    
    protected ArrayList<Treasure> hiddenTreasures = new ArrayList();
    protected ArrayList<Treasure> visibleTreasures = new ArrayList();
    
    public Player(String name){
        this.name = name;
        level = INITIALLEVEL;
        dead = true;
        canISteal = true;
        enemy = null;
        pendingBadConsequence = null;
    }
    
    public String getName(){ return name; }
    
    public String toString() {
        String s = this.getName() + "\nNivel de combate: " + this.getCombatLevel() + "\tNivel actual: " + this.getLevels() + "\n";
        return s;
    }
    
    public void checkPending() {
        System.out.println(this.pendingBadConsequence.toString());
    }
    
    private void bringToLife(){ dead = false;  }
    
    public int getCombatLevel(){
        int total = this.level;
        for(int i = 0; i < this.visibleTreasures.size(); i++){
            total += this.visibleTreasures.get(i).getBonus();
        }
        return total;
    }
    
    public int getLevels(){ return this.level; }    
    private void incrementLevels(int l){ 
        if(level + l > 10)
            level = 10;
        else
            level += l;
            }
    private void decrementLevels(int l){
        if(l >= level)
            level = 1;
        else
            level -= l;
            }
    private void setPendingBadconsequence(BadConsequence b){ pendingBadConsequence = b;} 
    
    public BadConsequence getPendingBadConsequence() {return this.pendingBadConsequence;}
    
    private void applyPrize(Monster monster){ 
        incrementLevels(monster.getLevelsGained());
        int nTreasures = monster.getTreasuresGained();
        
        if(nTreasures > 0){
         CardDealer dealer = CardDealer.getInstance();
         for(int i = 0; i < nTreasures; i++)
            hiddenTreasures.add(dealer.nextTreasure());
        }
    }
    
    
    
    private void applyBadConsequence(Monster monster){
       decrementLevels(monster.getBadConsequence().getLevels());
       
       BadConsequence b = monster.getBadConsequence();
       
       BadConsequence pendingBad = b.adjustToFitTreasureLists(this.visibleTreasures, this.hiddenTreasures);
       setPendingBadconsequence(pendingBad);
      
    }
    
    private boolean canMakeTreasureVisible(Treasure t){
       
        if(t.getType() != TreasureKind.ONEHAND && t.getType() != TreasureKind.BOTHHANDS){
            for (Treasure trea: visibleTreasures)
                if(trea.getType() == t.getType())
                    return false;
            return true;
        }
        else{
            int mano = 0;
            int manos = 0;
            for (Treasure trea: visibleTreasures){
                if(trea.getType() == TreasureKind.ONEHAND)
                  mano++;
                if(trea.getType() == TreasureKind.BOTHHANDS)
                  manos++;
            }
         if(mano == 0 && manos == 0)
             return true;
         
        return mano == 1 && t.getType() == TreasureKind.ONEHAND && manos == 0;
       
        }              
        
    }
    
    private int howManyVisibleTreasures(TreasureKind tkind){
        int total = 0;
        for(int i = 0; i < visibleTreasures.size(); i++){
            if(visibleTreasures.get(i).getType() == tkind)
                total +=  1;
        }
        return total;
    }
    
    private void dieIfNoTreasures(){
        if (hiddenTreasures.isEmpty() && visibleTreasures.isEmpty()) 
            dead = true;
    }

    public boolean isDead(){ return dead; }    
    public ArrayList<Treasure> getHiddenTreasures(){ return hiddenTreasures; }    
    public ArrayList<Treasure> getVisibleTreasures(){ return visibleTreasures; }

    
    public CombatResult combat(Monster monster){
        int myLevel = getCombatLevel();
        int monsterLevel = monster.getCombatLevel();
        if(!canISteal){
            Dice dice = Dice.getInstance();
            int number = dice.nextNumber();
            if(number < 3)
                monsterLevel += enemy.getCombatLevel();            
        }
        if(myLevel > monsterLevel){
            applyPrize(monster);
            if(level >= MAXLEVEL)
                return CombatResult.WINGAME;
            return CombatResult.WIN;
        }
        else{
            applyBadConsequence(monster);
            if (shouldConvert())
                return CombatResult.LOSEANDCONVERT;
            return CombatResult.LOSE;
        }
        
    }
    
    private boolean shouldConvert(){
        return (int) (Math.random()*6) + 1 == 6;
    }
    
    public void makeTreasureVisible(Treasure t){
        boolean canI = canMakeTreasureVisible(t);
        if(canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t); 
        if((pendingBadConsequence!=null) && (!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractVisibleTreasure(t);
        
            if((pendingBadConsequence.getNVisibleTreasures() == 0) && (pendingBadConsequence.getNHiddenTreasures() == 0)){
                pendingBadConsequence = null;
            }
        }
        dieIfNoTreasures();        
    }
    
    
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t); 
        if((pendingBadConsequence!=null) && (!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractHiddenTreasure(t);
        
            if((pendingBadConsequence.getNVisibleTreasures() == 0) && (pendingBadConsequence.getNHiddenTreasures() == 0)){
                pendingBadConsequence = null;
            }
        }
        dieIfNoTreasures();      
    }
    
    
    public boolean validState(){ 
        if(pendingBadConsequence== null )
            return true;
        else
            return pendingBadConsequence.isEmpty() && hiddenTreasures.size() < 5; 
    }
    
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        bringToLife();
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        int number = dice.nextNumber();
        if(number > 1){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
        if(number == 6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }      
                
    }
    
    
    public Treasure stealTreasure(){
        
        if(canISteal()){
            if(enemy.canYouGiveMeATreasure()) {
                Treasure treasure = enemy.giveMeATreasure();
                hiddenTreasures.add(treasure);
                haveStolen();
                return treasure;
            }
        }
        return null;
    }
    
    
    public void setEnemy(Player enemy){ this.enemy = enemy; }
    
    public Player getEnemy(){return this.enemy;}
    

    private Treasure giveMeATreasure(){
        int i;
        i = (int) (Math.random() * (hiddenTreasures.size()));
        
       return hiddenTreasures.remove(i); /*remove(índice) devuelve el objeto eliminadao*/
    }

    
    public boolean canISteal(){return canISteal; }    
    private boolean canYouGiveMeATreasure(){ return (hiddenTreasures.size() > 0); }
    
    
    private void haveStolen(){ canISteal = false; }
    
    
    public void discardAllTreasures(){
        ArrayList<Treasure> hcopy = new ArrayList(hiddenTreasures);
        ArrayList<Treasure> vcopy = new ArrayList(visibleTreasures);

        for (Treasure trea: vcopy)
            discardVisibleTreasure(trea);
        for (Treasure trea: hcopy)
            discardHiddenTreasure(trea);
    }
    
    
}





